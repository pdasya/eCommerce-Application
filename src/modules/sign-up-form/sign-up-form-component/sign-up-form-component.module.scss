.signUpForm {
  max-width: 400px;
  padding: 20px;
  border: 2px solid black;
  margin: 20px auto;
}

.submitButton {
  margin-top: 20px !important;
}

.dateOfBirthInput {
  color: red;
}

// input[type='date']:focus::-webkit-datetime-edit {
//   color: black !important;
// }

// input[type='date']:required:invalid::-webkit-datetime-edit {
//   color: transparent;
// }
