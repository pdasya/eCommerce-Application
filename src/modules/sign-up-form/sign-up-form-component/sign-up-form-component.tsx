import React, { FC } from 'react';
import { Formik, Form } from 'formik';
import { Button, Paper, Typography, Grid, Divider } from '@mui/material';
import * as Yup from 'yup';
import styles from './sign-up-form-component.module.scss';
import FieldComponent from '../components/field-component/field-component';
import { CustomRouterLink } from '@/components/custom-router-link/custom-router-link.component';
import { RoutePath } from '@/routes';

export interface IFormField {
  id: string;
  name: string;
  label: string;
  type: string;
  required?: boolean;
  options?: { label: string; value: string }[]; // For select input
}

interface ISignUpFormComponentProperties {
  title: string;
  buttonText: string;
  fields: IFormField[];
  initialValues: Record<string, string | boolean>;
  onSubmit: (values: Record<string, string | boolean>) => void;
}

const LinkToLoginPage: FC = () => (
  <CustomRouterLink to={RoutePath.signIn} className={styles.loginLink}>
    Already have an account? Login
  </CustomRouterLink>
);

export const getValidationSchema = (values: Record<string, string | boolean>) => {
  const minPasswordLength = 8;
  const minStreetNameLength = 1;
  const minAge = 13;

  const baseSchema = Yup.object().shape({
    email: Yup.string().email('Invalid email address').required('Email is required'),
    password: Yup.string()
      .min(minPasswordLength, `Password must be at least ${minPasswordLength} characters long`)
      .matches(
        /^(?=.*[a-z])(?=.*[A-Z])(?=.*\d).+$/,
        'Password must contain at least one uppercase letter, one lowercase letter, and one number',
      )
      .required('Password is required'),
    firstName: Yup.string()
      .matches(/^[A-Za-z]+$/, 'First name must contain only alphabetic characters')
      .required('First name is required'),
    lastName: Yup.string()
      .matches(/^[A-Za-z]+$/, 'Last name must contain only alphabetic characters')
      .required('Last name is required'),
    dateOfBirth: Yup.date()
      .max(
        new Date(new Date().setFullYear(new Date().getFullYear() - minAge)),
        `You must be at least ${minAge} years old`,
      )
      .typeError('Date of birth is required')
      .required('Date of birth is required'),
    shippingStreet: Yup.string()
      .min(minStreetNameLength, 'Street must contain at least one character')
      .required('Street is required'),
    shippingCity: Yup.string()
      .matches(/^[A-Za-z]+$/, 'City must contain only alphabetic characters')
      .required('City is required'),
    shippingPostalCode: Yup.string()
      .matches(/^\d{5}(-\d{4})?$/, 'Postal code must be in the format 12345 or 12345-6789')
      .required('Postal code is required'),
    shippingCountry: Yup.string()
      .oneOf(['US'], 'Invalid country selection')
      .required('Country is required'),
  });

  const billingSchema = Yup.object().shape({
    billingStreet: Yup.string()
      .min(minStreetNameLength, 'Street must contain at least one character')
      .required('Billing street is required'),
    billingCity: Yup.string()
      .matches(/^[A-Za-z]+$/, 'City must contain only alphabetic characters')
      .required('Billing city is required'),
    billingPostalCode: Yup.string()
      .matches(/^\d{5}(-\d{4})?$/, 'Postal code must be in the format 12345 or 12345-6789')
      .required('Billing postal code is required'),
    billingCountry: Yup.string()
      .oneOf(['US'], 'Invalid country selection')
      .required('Billing country is required'),
  });

  return values.setSameBillingAddress ? baseSchema : baseSchema.concat(billingSchema);
};

const validate = (values: Record<string, string | boolean>) => {
  try {
    getValidationSchema(values).validateSync(values, { abortEarly: false });
    return {};
  } catch (err) {
    const errors: Record<string, string> = {};
    if (err.inner) {
      err.inner.forEach((error: Yup.ValidationError) => {
        if (error.path) {
          errors[error.path] = error.message;
        }
      });
    }
    return errors;
  }
};

export const SignUpFormComponent: FC<ISignUpFormComponentProperties> = ({
  title,
  buttonText,
  fields,
  initialValues,
  onSubmit,
}) => (
  <Paper elevation={3} className={styles.signUpForm}>
    <Typography variant="h2" align="center" gutterBottom>
      {title}
    </Typography>
    <Formik
      initialValues={initialValues}
      validate={validate}
      onSubmit={async (values, { resetForm }) => {
        await onSubmit(values);
        resetForm();
      }}>
      {({ errors, touched, values }) => (
        <Form noValidate>
          <Grid container spacing={2}>
            {fields.map(field => {
              if (field.name === 'shippingStreet') {
                return (
                  <React.Fragment key={field.id}>
                    <Grid item xs={12}>
                      <Typography variant="subtitle1" className={styles.shippingAddressHeader}>
                        Shipping Address
                      </Typography>
                      <Divider className={styles.shippingAddressDivider} />
                    </Grid>
                    <Grid item xs={12} sm={6}>
                      <FieldComponent
                        field={field}
                        error={errors[field.name]}
                        touched={touched[field.name]}
                        values={values}
                      />
                    </Grid>
                  </React.Fragment>
                );
              }
              if (
                (field.name.includes('billing') || field.name.includes('DefaultBilling')) &&
                values.setSameBillingAddress
              ) {
                return null;
              }
              if (field.name === 'billingStreet') {
                return (
                  <React.Fragment key={field.id}>
                    <Grid item xs={12}>
                      <Typography variant="subtitle1" className={styles.billingAddressHeader}>
                        Billing Address
                      </Typography>
                      <Divider className={styles.billingAddressDivider} />
                    </Grid>
                    <Grid item xs={12} sm={6}>
                      <FieldComponent
                        field={field}
                        error={errors[field.name]}
                        touched={touched[field.name]}
                        values={values}
                      />
                    </Grid>
                  </React.Fragment>
                );
              }
              if (field.name === 'email') {
                return (
                  <React.Fragment key={field.id}>
                    <Grid item xs={12}>
                      <Typography variant="subtitle1" className={styles.generalInfoHeader}>
                        General Information
                      </Typography>
                      <Divider className={styles.generalInfoDivider} />
                    </Grid>
                    <Grid item xs={12} sm={6}>
                      <FieldComponent
                        field={field}
                        error={errors[field.name]}
                        touched={touched[field.name]}
                        values={values}
                      />
                    </Grid>
                  </React.Fragment>
                );
              }
              return (
                <Grid item xs={12} sm={6} key={field.id}>
                  <FieldComponent
                    field={field}
                    error={errors[field.name]}
                    touched={touched[field.name]}
                    values={values}
                  />
                </Grid>
              );
            })}
          </Grid>
          <Button
            type="submit"
            variant="contained"
            color="primary"
            fullWidth
            className={styles.submitButton}>
            {buttonText}
          </Button>
          <LinkToLoginPage />
        </Form>
      )}
    </Formik>
  </Paper>
);

export default SignUpFormComponent;
