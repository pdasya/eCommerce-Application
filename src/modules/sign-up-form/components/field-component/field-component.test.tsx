import React from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import { Formik } from 'formik';
import { IFormField } from '@modules/sign-up-form/sign-up-form-component/sign-up-form-component';
import userEvent from '@testing-library/user-event';
import FieldComponent from './field-component';
import '@testing-library/jest-dom';

const TestForm = ({
  fieldProps,
  initialValues,
  onSubmit,
}: {
  fieldProps: IFormField;
  initialValues: Record<string, string | boolean>;
  onSubmit: (values: Record<string, string | boolean>) => void;
}) => (
  <Formik initialValues={initialValues} onSubmit={onSubmit}>
    {() => <FieldComponent field={fieldProps} />}
  </Formik>
);

describe('FieldComponent', () => {
  const onSubmit = jest.fn();

  it('renders a select field and handles selection', async () => {
    const fieldProps = {
      id: 'country',
      name: 'country',
      label: 'Country',
      type: 'select',
      options: [
        { label: 'USA', value: 'USA' },
        { label: 'Canada', value: 'Canada' },
      ],
      required: true,
    };
    const initialValues = { country: '' };

    render(<TestForm fieldProps={fieldProps} initialValues={initialValues} onSubmit={onSubmit} />);
    const select = screen.getByRole('combobox');
    userEvent.click(select);

    const optionUSA = await screen.findByRole('option', { name: 'USA' });
    userEvent.click(optionUSA);

    await waitFor(() => {
      expect(optionUSA).toHaveAttribute('aria-selected', 'true');
    });
  });

  // it('renders a password field and validates input', async () => {
  //   const fieldProps = {
  //     id: 'password',
  //     name: 'password',
  //     label: 'Password',
  //     type: 'password',
  //     required: true
  //   };
  //   const initialValues = { password: '' };

  //   render(<TestForm fieldProps={fieldProps} initialValues={initialValues} onSubmit={onSubmit} />);

  //   const input = screen.getByLabelText('Password');
  //   await userEvent.type(input, 'password123');
  //   expect(input).toHaveValue('password123');
  // });

  it('renders a switch and toggles its state', async () => {
    const fieldProps = {
      id: 'newsletter',
      name: 'newsletter',
      label: 'Subscribe to newsletter',
      type: 'switch',
    };
    const initialValues = { newsletter: false };

    render(<TestForm fieldProps={fieldProps} initialValues={initialValues} onSubmit={onSubmit} />);
    const toggle = screen.getByRole('checkbox');
    await userEvent.click(toggle);
    expect(toggle).toBeChecked();
  });

  it('renders a checkbox and toggles its state', async () => {
    const fieldProps = {
      id: 'agree',
      name: 'agree',
      label: 'I agree to the terms',
      type: 'checkbox',
    };
    const initialValues = { agree: false };

    render(<TestForm fieldProps={fieldProps} initialValues={initialValues} onSubmit={onSubmit} />);
    const checkbox = screen.getByRole('checkbox');
    await userEvent.click(checkbox);
    expect(checkbox).toBeChecked();
  });

  // it('displays an error when the field is touched but invalid', async () => {
  //   const fieldProps = {
  //     id: 'email',
  //     name: 'email',
  //     label: 'Email',
  //     type: 'text',
  //     required: true,
  //   };
  //   const initialValues = { email: '' };

  //   render(<TestForm fieldProps={fieldProps} initialValues={initialValues} onSubmit={onSubmit} />);
  //   const emailInput = screen.getByRole('textbox', { name: /Email/i });
  //   await userEvent.type(emailInput, 'invalidemail');
  //   await userEvent.tab();
  //   await waitFor(() => expect(screen.getByText('Invalid email address')).toBeInTheDocument());
  // });
  it('matches the component snapshot', () => {
    const fieldProps = {
      id: 'testInput',
      name: 'testInput',
      label: 'Test Input',
      type: 'text',
      required: true,
    };
    const initialValues = { testInput: '' };

    const { asFragment } = render(
      <TestForm fieldProps={fieldProps} initialValues={initialValues} onSubmit={onSubmit} />,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
