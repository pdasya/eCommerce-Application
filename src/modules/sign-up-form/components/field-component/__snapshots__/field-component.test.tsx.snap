// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FieldComponent matches the component snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root"
  >
    <div
      class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-standard MuiFormLabel-colorPrimary Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-standard css-aqpgxn-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="false"
        for=":r1:"
        id=":r1:-label"
      >
        Test Input
        <span
          aria-hidden="true"
          class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
        >
          â€‰*
        </span>
      </label>
      <div
        class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-953pxc-MuiInputBase-root-MuiInput-root"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
          id=":r1:"
          name="testInput"
          required=""
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
